nixos/modules/system/etc/setup-etc.pl

```diff
@@ -85,62 +94,78 @@ sub cleanup {
 my %created;
 my @copied;
 
-sub link {
-    my $fn = substr $File::Find::name, length($etc) + 1 or next;
+sub make_symlinks {
+    my $path_to_file = $_;
+    # This is needed to avoid a warning if we take the substr of the first
+    # directory, i.e., the /nix/store/.../etc itself
+    if (length($File::Find::name) == length($etc)) {
+        return;
+    }
+
+    # Strip away /nix/store/<hash>-etc/etc/ from filename
+    my $fn = substr($File::Find::name, length($etc) + 1) or return;
     my $target = "/etc/$fn";
-    File::Path::make_path(dirname $target);
+    File::Path::make_path(dirname($target));
     $created{$fn} = 1;
 
     # Rename doesn't work if target is directory.
-    if (-l $_ && -d $target) {
-        if (isStatic $target) {
-            rmtree $target or warn;
+    if (-l $path_to_file && -d $target) {
+        if (is_static($target)) {
+            rmtree($target) or warn("Failed to remove $target");
         } else {
-            warn "$target directory contains user files. Symlinking may fail.";
+            warn("$target directory contains user files. Symlinking may fail.");
         }
     }
 
-    if (-e "$_.mode") {
-        my $mode = read_file("$_.mode"); chomp $mode;
+    # This will set uid/gid if the options were set in the nix config
+    if (-e "$path_to_file.mode") {
+        chomp(my $mode = read_file("$path_to_file.mode"));
         if ($mode eq "direct-symlink") {
-            atomicSymlink readlink("$static/$fn"), $target or warn;
+            atomic_symlink(readlink("$static/$fn"), $target) or warn;
         } else {
-            my $uid = read_file("$_.uid"); chomp $uid;
-            my $gid = read_file("$_.gid"); chomp $gid;
-            copy "$static/$fn", "$target.tmp" or warn;
-            $uid = getpwnam $uid unless $uid =~ /^\+/;
-            $gid = getgrnam $gid unless $gid =~ /^\+/;
-            chown int($uid), int($gid), "$target.tmp" or warn;
-            chmod oct($mode), "$target.tmp" or warn;
-            rename "$target.tmp", $target or warn;
+            chomp(my $uid = read_file("$path_to_file.uid"));
+            chomp(my $gid = read_file("$path_to_file.gid"));
+            copy("$static/$fn", "$target.tmp") or warn;
```

Comment:
```suggestion
            copy("$static/$fn", "$target.tmp") or warn "Copy failed: $!";
```
Include error messages. Also, maybe this should be `die` instead?